src/core/p3m/influence_function.hpp

```diff
@@ -52,44 +53,43 @@
 template <std::size_t S, std::size_t m>
 double G_opt(int cao, double alpha, Utils::Vector3d const &k,
              Utils::Vector3d const &h) {
-  using namespace detail::FFT_indexing;
-  using Utils::int_pow;
-  using Utils::sinc;
-
-  auto constexpr two_pi = 2. * std::numbers::pi;
-  auto constexpr two_pi_i = 1. / two_pi;
-  auto constexpr limit = 30.;
 
   auto const k2 = k.norm2();
-  if (k2 == 0.0) {
-    return 0.0;
+  if (k2 == 0.) {
+    return 0.;
   }
 
-  double numerator = 0.0;
-  double denominator = 0.0;
-
-  for (int mx = -m; mx <= m; mx++) {
-    for (int my = -m; my <= m; my++) {
-      for (int mz = -m; mz <= m; mz++) {
-        auto const km =
-            k + two_pi * Utils::Vector3d{mx / h[RX], my / h[RY], mz / h[RZ]};
-        auto const U2 = std::pow(sinc(km[RX] * h[RX] * two_pi_i) *
-                                     sinc(km[RY] * h[RY] * two_pi_i) *
-                                     sinc(km[RZ] * h[RZ] * two_pi_i),
-                                 2 * cao);
-
+  auto constexpr limit = 30.;
+  auto constexpr m_start = Utils::Vector3i::broadcast(-m);
+  auto constexpr m_stop = Utils::Vector3i::broadcast(m + 1);
+  auto const exponent_prefactor = Utils::sqr(1. / (2. * alpha));
+  auto const wavevector = (2. * std::numbers::pi) / h;
+  auto const wavevector_i = 1. / wavevector;
+  auto indices = Utils::Vector3i{};
+  auto km = Utils::Vector3d{};
+  auto fnm = Utils::Vector3d{};
+  auto numerator = 0.;
+  auto denominator = 0.;
+
+  for_each_3d(
+      m_start, m_stop, indices,
+      [&]() {
+        using namespace detail::FFT_indexing;
+        auto const U2 = std::pow(Utils::product(fnm), 2 * cao);
```

Comment:
In all the influence functions we could save one multiplication per iteration by caching the products in `fnm`, such that `fnm[0] = std::pow(math::sinc(...), 2 * cao);` and `fnm[1] = fnm[0] + std::pow(...)`, ... I'm not sure if it's worth the effort, since we probably don't do this often enough in a simulation

Reply:
Could be. This specific diff already improves performance by caching the intermediate `sinc()` calls. But the P3M benchmarks reveal only 2% performance improvement for this entire PR. I gave up optimizing these 3D kernels, because saving a float multiplication is probably not worth it. It's also unclear, whether we will still be able to cache the `fnm` variable once we replace it with Kokkos parallel execution (OpenMP, potentially out-of-order).